### ISSUE TITLE
Protocol\Smtp->data() uses too much memory

### ISSUE THREAD

**Issue by echampet** [zendframework/zend-mail#8]
Should i just increase memory_limit to 256M (not always possible)
or do you guys are ok with loosing some raw performance?

With huge mails (50M, 660000 lines), i'm hitting memory_limit, so i've done some tests
looking at:
- peak memory usage for big mails (50M, 660000 lines)
- average time(100 runs) for big mails
- average time(1000 runs) for small mails (50k, 1700 lines)

with php5.6/php7 (on a fedora 22, core i7, ssd)

- str_replace("\r","") + explode("\n")
(what we use now)

260M/200M|0,23s/0,08s|0,29ms/0,18ms
```
$data = str_replace("\r", '', $data);
$data = explode("\n", $data);
foreach ($data as $line) {
if (isset($line[0]) && $line[0] === '.') {
    $line = '.' . $line;
}
//send
}
```

- preg_split("\r?\n")

200M/160M|1,5s/0,14s|1,75ms/0,22ms
```
foreach (preg_split('/\r?\n/', $data) as $line) {
if (isset($line[0]) && $line[0] === '.') {
    $line = '.' . $line;
}
//send
}
```

- strpos("\n") + $data[]==="\r" + substr()

50M/50M|0,32s/0,16s|0,86ms/0,41ms
```
$start = 0;
while (($stop = strpos($data, "\n", $start)) !== false) {
$len = $stop-$start;
if ($len > 0 && $data[$stop-1] === "\r") {
    $len -= 1;
}
$line = substr($data, $start, $len);

if (isset($line[0]) && $line[0] === '.') {
    $line = '.' . $line;
}
//send
$start = $stop + 1;
}
//last line
$line = substr($data, $start);
if (isset($line[0]) && $line[0] === '.') {
$line = '.' . $line;
}
//send
```

- str_replace("\r","") + strpos("\n") + substr()

150M/100M|0,28s/0,14s|0,76ms/0,31ms
```
$start = 0;
$data = str_replace("\r", '', $data);
while (($stop = strpos($data, "\n", $start)) !== false) {
$line = substr($data, $start, $stop-$start);
if (isset($line[0]) && $line[0] === '.') {
    $line = '.' . $line;
}
//send
$start = $stop + 1;
}
//last line
$line = substr($data, $start);
if (isset($line[0]) && $line[0] === '.') {
$line = '.' . $line;
}
//send
```

- fwrite(php://temp) + stream_get_line("\n") + rtrim("\r")

50M/50M|0,28s/0,18s|0,68ms/0,40ms
```
$fp = fopen("php://temp", "r+");
fwrite($fp, $data);
unset($data);
rewind($fp);
while (($line = stream_get_line($fp, 1000, "\n")) !== false) {
$line = rtrim($line, "\r");
if (isset($line[0]) && $line[0] === '.') {
    $line = '.' . $line;
}
//send
}
```

- strtok("\n") + rtrim("\r") (convert multiple \n\n\n in 1 -> ko)

All tests were run with 'php -d"memory_limit=256M" -d"opcache.enable_cli=1"'

**Comment by Maks3w** [comment]
Have you tried http://3v4l.org/ for profile your optimizations across different PHP versions?

**Comment by echampet** [comment]
no, thanks a lot, i'll try

**Comment by echampet** [comment]
hi all,
here the 2 interesting results (str_replace("\r","") + explode("\n") VS strpos("\n") + $data[]==="\r" + substr())
with 10000 lines:
http://3v4l.org/eiGNO/perf#tabs
http://3v4l.org/Uct3Q/perf#tabs
with 100000 lines:
http://3v4l.org/vfrZj/perf#tabs
http://3v4l.org/BK4tX/perf#tabs
and with 500000 lines (str_replace("\r","") + explode("\n") hit memory limit):
http://3v4l.org/7TP84/perf#tabs
http://3v4l.org/C2XRb/perf#tabs

strpos("\n") + $data[]==="\r" + substr() is a bit slower, but memory usage doesn't explode, thus allowing to handle big emails

Can someone  consider merging #6? (else please merge #4)

**Comment by champtar** [comment]
ping @Maks3w

**Comment by echampet** [comment]
ping @Maks3w

**Comment by Ocramius** [comment]
@echampet directly suggest a patch :)

**Comment by samsonasik** [comment]
should already fixed by #7 , close-able

**Comment by echampet** [comment]
yep
